app/models/build/locking.rb

```
@@ -0,0 +1,47 @@
+module Build
+  module Locking
+    extend self
+
+    # Create an exclusive lock and yield
+    def lock(lock_name, options = {})
+      mutex = build_mutex(lock_name, options)
+
+      begin
+        mutex.lock
```

I couldn't find any place where any locking exception is explicitly caught. Is there any reason to fail early? For example, in the main controller, I put the check if the component needs building inside of the lock, so any thread that gets into the block won't do unnecessary work.